modal-go/task_command_router_client.go

```diff
@@ -0,0 +1,629 @@
+	}, opts)
+
+	if err != nil {
+		st, ok := status.FromError(err)
+		if (ok && st.Code() == codes.DeadlineExceeded) || errors.Is(err, errDeadlineExceeded) {
+			return nil, fmt.Errorf("deadline exceeded while waiting for exec %s", execID)
+		}
+	}
+	return resp, err
+}
+
+// stdioReadResult represents a result from the stdio read stream.
+type stdioReadResult struct {
+	Response *pb.TaskExecStdioReadResponse
+	Err      error
+}
+
+// ExecStdioRead reads stdout or stderr from an exec.
+// The returned channel will be closed when the stream ends or an error occurs.
+func (c *TaskCommandRouterClient) ExecStdioRead(
+	ctx context.Context,
+	taskID, execID string,
+	fd pb.FileDescriptor,
+	deadline *time.Time,
+) <-chan stdioReadResult {
+	resultCh := make(chan stdioReadResult)
+
+	go func() {
+		defer close(resultCh)
+
+		var srFd pb.TaskExecStdioFileDescriptor
+		switch fd {
+		case pb.FileDescriptor_FILE_DESCRIPTOR_STDOUT:
+			srFd = pb.TaskExecStdioFileDescriptor_TASK_EXEC_STDIO_FILE_DESCRIPTOR_STDOUT
+		case pb.FileDescriptor_FILE_DESCRIPTOR_STDERR:
+			srFd = pb.TaskExecStdioFileDescriptor_TASK_EXEC_STDIO_FILE_DESCRIPTOR_STDERR
+		case pb.FileDescriptor_FILE_DESCRIPTOR_INFO, pb.FileDescriptor_FILE_DESCRIPTOR_UNSPECIFIED:
+			resultCh <- stdioReadResult{Err: fmt.Errorf("unsupported file descriptor: %v", fd)}
+			return
+		default:
+			resultCh <- stdioReadResult{Err: fmt.Errorf("invalid file descriptor: %v", fd)}
+			return
+		}
+
+		c.streamStdio(ctx, resultCh, taskID, execID, srFd, deadline)
+	}()
+
+	return resultCh
+}
+
+// MountDirectory mounts an image at a directory in the container.
+func (c *TaskCommandRouterClient) MountDirectory(ctx context.Context, request *pb.TaskMountDirectoryRequest) error {
+	_, err := callWithRetriesOnTransientErrors(ctx, func() (struct{}, error) {
+		callErr := c.callWithAuthRetry(ctx, func(authCtx context.Context) error {
+			_, err := c.stub.TaskMountDirectory(authCtx, request)
+			return err
+		})
+		return struct{}{}, callErr
+	}, defaultRetryOptions())
+	return err
+}
+
+// SnapshotDirectory snapshots a directory into a new image.
+func (c *TaskCommandRouterClient) SnapshotDirectory(ctx context.Context, request *pb.TaskSnapshotDirectoryRequest) (*pb.TaskSnapshotDirectoryResponse, error) {
+	resp, err := callWithRetriesOnTransientErrors(ctx, func() (*pb.TaskSnapshotDirectoryResponse, error) {
+		var resp *pb.TaskSnapshotDirectoryResponse
+		callErr := c.callWithAuthRetry(ctx, func(authCtx context.Context) error {
+			var err error
+			resp, err = c.stub.TaskSnapshotDirectory(authCtx, request)
+			return err
+		})
+		return resp, callErr
+	}, defaultRetryOptions())
+	return resp, err
+}
+
+func (c *TaskCommandRouterClient) streamStdio(
+	ctx context.Context,
+	resultCh chan<- stdioReadResult,
+	taskID, execID string,
+	fd pb.TaskExecStdioFileDescriptor,
+	deadline *time.Time,
+) {
+	if deadline != nil {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithDeadline(ctx, *deadline)
```

Comment:
Should the caller set `context.WithDeadline` such that `deadline` does not need to passed in here anymore?